source/beamline/SGM/SGMBeamCoordinatorControl.cpp

```diff
@@ -0,0 +1,252 @@
+	psh_2_ = new CLSBiStateControl("PSH-2", "PSH1411-I00-02", "PSH1411-I00-02:state", "PSH1411-I00-02:opr:open", "PSH1411-I00-02:opr:close", new AMControlStatusCheckerDefault(2), this, CHILD_TIMEOUT);
+
+	es2Bypass_ = new AMReadOnlyPVControl("ES2 Bypass", "SWZ1611-4-02:fbk",this);
+
+	addChildControl(fastShutterVoltage_);
+	addChildControl(frontBypassValve_);
+	addChildControl(backBypassValve_);
+
+	addChildControl(vvr1611_3_I10_01Valve_);
+	addChildControl(vvr1611_3_I10_02Valve_);
+	addChildControl(vvr1611_3_I10_03Valve_);
+	addChildControl(vvr1611_3_I10_04Valve_);
+	addChildControl(psh1611_3_I10_01Shutter_);
+	addChildControl(vvr1611_4_I10_01Valve_);
+	addChildControl(vvr1611_4_I10_02Valve_);
+	addChildControl(vvr1611_4_I10_03Valve_);
+	addChildControl(vvr1611_4_I10_04Valve_);
+	addChildControl(vvr1611_4_I10_05Valve_);
+	addChildControl(vvr1611_4_I10_06Valve_);
+	addChildControl(vvr1611_4_I10_07Valve_);
+	addChildControl(vvr1611_4_I10_08Valve_);
+	addChildControl(vvr1611_4_I10_11Valve_);
+	addChildControl(psh1411_I00_01Shutter_);
+	addChildControl(vvr1411_I00_01Valve_);
+	addChildControl(vvf1411_I00_01_);
+	addChildControl(safetyShutter_);
+	addChildControl(vvr1611_3_I00_01Valve_);
+	addChildControl(vvr1611_3_I00_02Valve_);
+	addChildControl(es2Bypass_);
+
+	controlsToOpen_.append(vvr1611_3_I10_01Valve_);
+	controlsToOpen_.append(vvr1611_3_I10_01Valve_);
+	controlsToOpen_.append(vvr1611_3_I10_02Valve_);
+	controlsToOpen_.append(vvr1611_3_I10_03Valve_);
+	controlsToOpen_.append(vvr1611_3_I10_04Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_01Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_02Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_03Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_04Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_05Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_06Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_07Valve_);
+	controlsToOpen_.append(vvr1611_4_I10_08Valve_);
+	controlsToOpen_.append(frontBypassValve_);
+	controlsToOpen_.append(backBypassValve_);
+	controlsToOpen_.append(vvr1611_4_I10_11Valve_);
+	controlsToOpen_.append(vvr1611_3_I00_01Valve_);
+	controlsToOpen_.append(vvr1611_3_I00_02Valve_);
+
+
+	controlsToWait_.append(psh1411_I00_01Shutter_);
+	controlsToWait_.append(vvr1411_I00_01Valve_);
+	controlsToWait_.append(vvf1411_I00_01_);
+
+	setMinimumValue(0);
+	setMaximumValue(1);
+	setTolerance(0.5);
+
+	setAllowsMovesWhileMoving(false);
+	setAttemptMoveWhenWithinTolerance(false);
+	updateStates();
+}
+
+bool SGMBeamCoordinatorControl::canMeasure() const
+{
+	bool canMeasureResult = true;
+	foreach(AMControl* childControl, children_) {
+		canMeasureResult &= childControl->canMeasure();
+	}
+
+	return canMeasureResult;
+}
+
+bool SGMBeamCoordinatorControl::canMove() const
+{
+	return isConnected();
+}
+
+bool SGMBeamCoordinatorControl::canStop() const
+{
+	return false;
```

Comment:
I believe it could be possible with the following changes:
- Add a stop PV to the energy db file for the beam on/off actions.
- Wire up the stop PV in the coordinator to pass it along to the beam coordinator control
- Store a copy of the move actions in the beam coordinator control
- When the stop PV is triggered, cancel the list of actions in the beam coordinator control
- Add an override for onMoveCancelled to the coordinator control
- Within the onMoveCancelled, onMoveSucceeded and onMoveFailed function in the coordinator control set the stored action to 0
- Add a button to the beam on control widget.

I think that would be it.